mp4e/mp4.go

```diff
@@ -0,0 +1,308 @@
+	}
+	if file.Segments[0].Sidx != nil {
+		info.Timescale = file.Segments[0].Sidx.Timescale
+	}
+
+	dtsPrev := uint64(0)
+	ptsPrev := uint64(0)
+	sampleCountPrev := uint64(0)
+	sampleCountPrevPrev := uint64(0)
+	seqPrev := uint32(0)
+	for i, seg := range file.Segments {
+		info.FragmentCount += uint64(len(seg.Fragments))
+		segInfo := SegmentInfo{
+			FragmentCount: uint64(len(seg.Fragments)),
+		}
+
+		if seg.Fragments == nil {
+			info.AddError("No fragments", i, -1, -1)
+		}
+
+		// moof - movie fragment
+		//   mfhd - movie fragment header; sequence number
+		//   traf - track fragment
+		//     tfhd - track fragment header; track ID, default sample duration
+		//     tfdt - track fragment decode time
+		for j, frag := range seg.Fragments {
+			seq := uint32(0)
+
+			if frag.Moof == nil {
+				info.AddError("No moof", i, j, -1)
+			} else {
+				if frag.Moof.Mfhd == nil {
+					info.AddError("No mfhd", i, j, -1)
+				} else {
+					seq = frag.Moof.Mfhd.SequenceNumber
+				}
+
+				//if frag.Moof.Traf == nil {
+				//	info.AddError("No traf", i, j)
+				//} else {
+				//	if frag.Moof.Traf.Tfhd == nil {
+				//		info.AddError("No tfhd", i, j)
+				//	} else {
+				//		dur = uint64(frag.Moof.Traf.Tfhd.DefaultSampleDuration)
+				//		if dur == 0 {
+				//			info.AddError("default_sample_duration is 0", i, j)
+				//		}
+				//	}
+				//	if frag.Moof.Traf.Tfdt == nil {
+				//		info.AddError("No tfdt", i, j)
+				//	} else {
+				//		dts = frag.Moof.Traf.Tfdt.BaseMediaDecodeTime()
+				//	}
+				//}
+			}
+			//if len(seg.Sidxs) > j {
+			//	sidx := seg.Sidxs[j]
+			//	pts = sidx.EarliestPresentationTime
+			//}
+
+			if j == 0 {
+				segInfo.SeqStart = seq
+			} else if j == len(seg.Fragments)-1 {
+				segInfo.SeqEnd = seq
+			}
+
+			if seq != seqPrev+1 && j > 0 {
+				info.AddError(fmt.Sprintf("Sequence number gap %d - %d", seqPrev, seq), i, j, -1)
+			}
+			seqPrev = seq
+
+			var samples []mp4.FullSample
+			samples, err = frag.GetFullSamples(nil)
+			if err != nil {
+				info.AddError(fmt.Sprintf("failed to read samples: %v", err), i, j, -1)
+				continue
+			}
+			sampleCount := uint64(len(samples))
+			segInfo.SampleCount += sampleCount
+			for k, sample := range samples {
+				dur := uint64(sample.Dur)
+				if dur < info.SampleDurationMin || info.SampleDurationMin == 0 {
+					info.SampleDurationMin = dur
+				}
+				if dur > info.SampleDurationMax {
+					info.SampleDurationMax = dur
+				}
+				dts := sample.DecodeTime
+				pts := sample.PresentationTime()
+				segInfo.Samples = append(segInfo.Samples, &SampleInfo{dur, j, pts})
+				if j == 0 && k == 0 {
+					segInfo.DtsStart = dts
+					segInfo.PtsStart = pts
+					if err = ensureIDRFrame(sample); err != nil {
+						info.AddError(err.Error(), i, j, int64(pts))
+					}
+				} else if j == len(seg.Fragments)-1 && k == len(samples)-1 {
+					segInfo.DtsEnd = dts + dur
+					segInfo.PtsEnd = pts + dur
+				}
+				if dtsPrev != 0 && absDiff(dts-dtsPrev, dur) > 1 {
```

Comment:
So this allows a gap of 1 TS before raising an issue with it? Is that normal (I genuinely have no context here)?

Reply:
Also, does this fully account for cases where these are not delivered in order? Should this check not be made relative to PTS, after ordering by PTS?

Reply:
Uh yeah I don't remember why it's not 0, so it might just be a bug.

DTS should be in order I think (it makes sense to feed the decoder in order)